mgr/volumes: use depth-first, non-recursive approach for cloning

rishabh-d-dave opened the pull request:
Switch to non-recursive approach in cptree() (located in bulk_copy() in
async_cloner.py) to prevent it from crashing with "RecursionError:
maximum recursion depth exceeded".

Fixes: https://tracker.ceph.com/issues/72357






<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

rishabh-d-dave committed:
pybind/cephfs: use depth-first, non-recursive approach for cloning

Switch to non-recursive approach in cptree() (located in bulk_copy() in
async_cloner.py) to prevent it from crashing with "RecursionError:
maximum recursion depth exceeded".

Fixes: https://tracker.ceph.com/issues/72357
Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave commented:
jenkins retest this

batrick reviewed (changes requested):
Similarly to your other PR, I want to see `createat`, etc. going forward. It's a better way to structure this recursive operation.

rishabh-d-dave committed:
test_cephfs: add unit tests for cptree() in cephfs python bindings

Signed-off-by: Rishabh Dave <ridave@redhat.com>

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2878,6 +2878,297 @@ cdef class LibCephFS(object):
+            src_slink_name = os.path.basename(src_path)
+            dst_path = os.path.join(dst_path, src_slink_name)
+
+            copy_sym_link(self, src_path, dst_path, mode)
```
I don't understand special casing this here. The `CopyTreeWorker` should just handle that.

When you finally use `openat` style API, you can change your constructor to:

```
    def __init__(self, fs, srcdir: int, path1: str, destdir: int, dst_path2: str, should_cancel, suppress_errors=False):
```

rishabh-d-dave committed:
pybind/cephfs: add python binding for mkdirat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add python bindings for statxat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add python binding for readlinkat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: add test for readlinkat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add python binding for symlinkat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: add tests for symlinkat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add python bindings for chmodat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: add tests for chmodat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add tests for statxat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
mgr/volumes: clone using cptree() from cephfs python bindings

Fixes: https://tracker.ceph.com/issues/72357
Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add python bindings for chownat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: add python bindings for utimensat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
qa/cephfs: add tests for chownat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: add tests for utimensat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs: call object setup/teardown for all tests in TestWithRootUser

Signed-off-by: Rishabh Dave <ridave@redhat.com>

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -3133,3 +3430,334 @@ class RmtreeDir:
+    def notify_parent_dir(self):
+        '''
+        Add current dir's name to parent dir's "de_ignore_list". This is
+        necessary since parent dir can't be deleted when current dir can't be
```
? copy paste error?

vshankar commented:
Tagging @joscollin @neesingh-rh for review as @batrick will not be able to get to it anytime soon (mentioned to @rishabh-d-dave over slack).

In the meantime, pleas rebase this @rishabh-d-dave and incorporate any review comments.

joscollin commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -8998,7 +8946,7 @@ def test_subvolume_snapshot_reconf_max_concurrent_clones(self):
         self.assertEqual(max_concurrent_clones, 2)
 
-    def test_subvolume_snapshot_config_snapshot_clone_delay(self):
+    def _test_subvolume_snapshot_config_snapshot_clone_delay(self):
```
Please drop this. You could push this to your origin for testing instead.

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -864,6 +864,32 @@ def test_multi_target_command():
+    Tests for libcephfs's fdopendir().
+    '''
+
+    def test_fdopendir_for_dir_at_CWD(self, testdir):
```
`testdir` is not used anywhere

rishabh-d-dave replied:
No, `testdir` is a pytest fixture.

joscollin replied:
okay, but does it require a `@pytest.fixture` decorator?

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -864,6 +864,32 @@ def test_multi_target_command():
+        cephfs.mkdir(dirname, 0o755)
+
+        fd = cephfs.open(dirname, os.O_RDONLY | os.O_DIRECTORY, 0o755)
+        dh = cephfs.fdopendir(fd)
```
Check if `dh` is valid here and if not then fail the test?

rishabh-d-dave replied:
Not needed, IMO as a failure/crash wouldn't automatically tell the reason

joscollin commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -1156,6 +1156,25 @@ cdef class LibCephFS(object):
+
+        if ret < 0:
+            raise make_ex(ret, f"error in mkdirat: {relpath.decode('utf-8')}")
+
```
No test that uses mkdirat() ?

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+        cephfs.cptree(slink, dst, should_cancel=should_cancel,
+                      suppress_errors=False)
+
+        cephfs.stat(f'{dst}/{slink}', follow_symlink=False)
```
So it checks that copying slink is possible when there's no permission?

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+        cephfs.stat(dst)
+        for i in range(1, 6):
+            cephfs.stat(f'{src}/file{i}')
+            cephfs.stat(f'{dst}/{src}/file{i}')
```
Just compare checksum.

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+    def test_cptree_when_tree_contains_dirs_regfiles_and_symlinks(self, testdir):
+        '''
+        Test that cptree() successfully copies entire file hierarchy that
+        contains directories, regular files as well as symbolic links.
```
This test checks dirs, regfiles and symlinks. So can we drop some of the previous test functions that involves checking dirs/regfiles/symlinks?

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+        '''
+        Test that when should_sync_attrs=True for cptree(), attributes
+        specifically uid, gid, owner, mode, atime and mtime) are synced
+        on dst files to match src files.
```
Fix comment. `should_sync_attrs` is not True here

rishabh-d-dave replied:
Fixed.

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+        cephfs.mkdir(src, 0o755)
+        cephfs.mkdir(dst, 0o755)
+
+        for i in range(1, 201):
```
not 2k

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -917,6 +917,52 @@ def test_openat_for_grandchild_subdir(self, testdir):
+        fd = cephfs.open('dir1', os.O_RDONLY | os.O_DIRECTORY, 0o755)
+        # bug? libcephfs.AT_SYMLINK_NOFOLLOW has no effect.
+        statx_buf = cephfs.statxat(fd, b'file1', libcephfs.CEPH_STATX_MODE,
+                                   libcephfs.AT_SYMLINK_NOFOLLOW)
```
statxat `slink1` right? Not `file1`.
Or have I misunderstood the use of `libcephfs.AT_SYMLINK_NOFOLLOW`?

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -304,6 +304,27 @@ def test_symlink(testdir):
+    cephfs.close(fd)
+
+    fd = cephfs.open('dir1', os.O_RDONLY | os.O_DIRECTORY, 0o755)
+    cephfs.symlinkat('file1', fd, 'slink1')
```
What's the relevance of `slink1` in this test? The test seem to pass without it. Could you please clarify?

joscollin commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2814,6 +2814,43 @@ cdef class LibCephFS(object):
+            ret = ceph_utimensat(self.cluster, _fd, _relpath, _times, _flags)
+
+        if ret < 0:
+            raise make_ex(ret, "error in utimensat")
```
`error setting access and modification time`

rishabh-d-dave commented on src/test/pybind/test_cephfs.py:
```diff
@@ -864,6 +864,32 @@ def test_multi_target_command():
+        cephfs.mkdir(dirname, 0o755)
+
+        fd = cephfs.open(dirname, os.O_RDONLY | os.O_DIRECTORY, 0o755)
+        dh = cephfs.fdopendir(fd)
```
same as https://github.com/ceph/ceph/pull/64774/files#r2763026807.

rishabh-d-dave commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+    def test_cptree_on_a_200_dir_broad_tree(self, testdir):
+        '''
+        Test that cptree() successfully copies a file hierarchy with 200
+        subdirectories on the same level.
```
We saw some issues with 2k levels of dir previously but not with 200, that's why we have separate tests for both.

rishabh-d-dave commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+    def test_cptree_on_a_200_dir_deep_tree(self, testdir):
+        '''
+        Test that cptree() successfully copies a file hierarchy with 200
+        levels.
```
same as https://github.com/ceph/ceph/pull/64774/files#r2763049430.

rishabh-d-dave commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1048,6 +1048,56 @@ def _test_chmodat_on_link(self, testdir):
+        fd = cephfs.open('dir1/file1', 'w', 0o644)
+        cephfs.write(fd, b'abcd', 0)
+        cephfs.close(fd)
+        cephfs.symlink('file1', 'dir1/slink1')
```
Current arguments are correct, they have the exact effect I want them to have.

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1929,3 +1930,884 @@ def test_rmtree_on_a_very_very_deep_tree(self, testdir):
+        for i in range(1, 6):
+            cephfs.stat(f'{dst}/{src}/dir11/file{i}')
+            # ensure src was left as it was.
+            cephfs.stat(f'{src}/dir11/file{i}')
```
Right, but you need to check `snap1`s absence in `dst`. That's the goal of this test, right?

joscollin commented on src/test/pybind/test_cephfs.py:
```diff
@@ -1785,118 +1930,885 @@ def test_rmtree_on_a_2k_dir_deep_tree(self, testdir):
+            # ensure src was left as it was.
+            cephfs.stat(f'{src}/dir11/file{i}')
+
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, '/dir2/dir1/dir11/.snap/snap1')
```
Change it to `f'{dst}/{src}/dir11/.snap/snap1`

rishabh-d-dave commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -3124,6 +3125,56 @@ cdef class LibCephFS(object):
+        if isinstance(dst_path, str):
+            dst_path = dst_path.encode('utf-8')
+        else:
+            assert isinstance(src_path, bytes)
```
> assert isinstance(dst_path, str)

This is wrong, bytes object is also acceptable.

rishabh-d-dave committed:
test_cephfs.py: add tests for mkdirat()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave commented:
.

rishabh-d-dave committed:
test/pybind/assertions: add helper method assert_less

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: make should_cancel option parameter for rmtree()

Make it optional to pass a function reference for should_cancel
parameter of rmtree function by defining a default value for it. Right
now one has to pass a simple lambda function returning False if one
wants to make rmtree operation uninterruptible -

rmtree('some-dir', should_cancel=lambda: False)

With should_cancel's default value set to "lambda: False", the call is
much simplied -

rmtree('some-dir')

Fixes: https://tracker.ceph.com/issues/75926
Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
pybind/cephfs: don't attempt to unlink root in rmtree

Don't attempt to unlink '/' as part of recursive delete when rmtree('/')
is called as it is an invalid op.

Fixes: https://tracker.ceph.com/issues/75924
Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: delete purge_dir() helper method, use rmtree() instead

Use rmtree() instead of purge_dir() for following 2 reasons -

1. purge_dir()'s recursive nature makes it vulnerable to max recursion
   limit.
2. It is redundant to have purge_dir() helper method since rmtree()
   already performs job in a better way.

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: test rmtree on root

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave committed:
test_cephfs.py: test rmtree with and without should_cancel

Signed-off-by: Rishabh Dave <ridave@redhat.com>

joscollin commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -3379,3 +3430,335 @@ class RmtreeDir:
+                    self.is_empty = False
+                    return de
+
+                de = self.fs.readdir(self.src_handle)
```
@rishabh-d-dave This is not done

joscollin commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -1419,6 +1419,40 @@ cdef class LibCephFS(object):
+        :param gid: the gid to set
+        :param follow_symlink: perform the operation on the target file if @path
+                               is a symbolic link (default)
+        """
```
s/flag/flags

rishabh-d-dave committed:
test_cephfs.py: remove rendundant call to purge_dir()

Signed-off-by: Rishabh Dave <ridave@redhat.com>

rishabh-d-dave commented:
.

rishabh-d-dave commented:
@joscollin See -

* https://github.com/ceph/ceph/pull/64774/changes#diff-6b5d3d15c9af71de126ede1bed024879f233e84663ccd4ff0b862b0869734d26R3507
* https://github.com/ceph/ceph/pull/64774/changes#diff-6b5d3d15c9af71de126ede1bed024879f233e84663ccd4ff0b862b0869734d26R3479
* https://github.com/ceph/ceph/pull/64774/changes#diff-82975fc360fb5539f1c93fede980b67a1a5cd5e0c81787ca907f02f4531a2bb0R2373
* https://github.com/ceph/ceph/pull/64774/changes#diff-82975fc360fb5539f1c93fede980b67a1a5cd5e0c81787ca907f02f4531a2bb0R2507
* https://github.com/ceph/ceph/pull/64774/changes#diff-82975fc360fb5539f1c93fede980b67a1a5cd5e0c81787ca907f02f4531a2bb0R2525

joscollin reviewed (approved):
Looks good to me.

joscollin commented:
jenkins render docs

joscollin commented:
jenkins test docs

rishabh-d-dave commented:
@vshankar Would you like to review and approve this PR or can we go ahead with QA for it? I ran libcephfs unit tests, those ran fine.

rishabh-d-dave commented:
Rebased on latest main.

vshankar commented:
> @vshankar Would you like to review and approve this PR or can we go ahead with QA for it? I ran libcephfs unit tests, those ran fine.

If this change follows closely with the rmtree change, then go ahead with QA.

joscollin commented:
> > @vshankar Would you like to review and approve this PR or can we go ahead with QA for it? I ran libcephfs unit tests, those ran fine.
> 
> If this change follows closely with the rmtree change, then go ahead with QA.

+1

rishabh-d-dave commented:
libcephfs test suite passed - https://pulpito.ceph.com/rishabh-2026-04-16_08:38:04-fs:libcephfs-rishabh-cptree-testing-default-trial/

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test windows

rishabh-d-dave commented:
jenkins test docs

vshankar commented:
This PR is under test in [https://tracker.ceph.com/issues/76232](https://tracker.ceph.com/issues/76232).

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/QA_main_--_2026#wip-vshankar-testing-20260504140938
